File: J9Ch/src/S_1_Follow_up_in_CodeInterview/Required_5/_386_Longest_Substring_with_At_Most_K_Distinct_Characters_Medium.py


class Solution:
    # @param s : A string
    # @return : An integer
    def lengthOfLongestSubstringKDistinct(self, s, k):
        if k == 0:
            return 0

        hash = {}
        head, tail = 0, 0
        longest = 0
        while head < len(s):
            while head < len(s) and (s[head] in hash or len(hash) < k):
                hash.setdefault(s[head], 0)
                hash[s[head]] += 1
                head += 1
            longest = max(longest, head - tail)
            while tail < head and len(hash) == k:
                hash[s[tail]] -= 1
                if hash[s[tail]] == 0:
                    del hash[s[tail]]
                tail += 1
        return longest

File: J9Ch/src/S_1_Follow_up_in_CodeInterview/Required_5/python_cpp/_386_Longest_Substring_with_At_Most_K_Distinct_Characters_Medium.cpp



class Solution {
public:
    /**
     * @param s : A string
     * @return : The length of the longest substring
     *           that contains at most k distinct characters.
     */
    int lengthOfLongestSubstringKDistinct(string s, int k) {
        // write your code here
        int start = 0, cnt = 0;
        int char_set[256] = {0};
        int ans = 0;
        for (int i = 0; i < s.size(); i++) {
            if (char_set[s[i]]++ == 0) cnt++;
            while (cnt > k) {
                char_set[s[start]]--;
                if (char_set[s[start++]] == 0) cnt--;
            }
            ans = max(i - start + 1, ans);
        }
        return ans;
    }
};